pandas/core/reshape/concat.py

```diff
@@ -516,9 +516,12 @@ def _get_new_axes(self) -> List[Index]:
 
     def _get_comb_axis(self, i: int) -> Index:
         data_axis = self.objs[0]._get_block_manager_axis(i)
-        return get_objs_combined_axis(
+        comb_axis = get_objs_combined_axis(
             self.objs, axis=data_axis, intersect=self.intersect, sort=self.sort
         )
+        # GH 29879
+        # No need for deep copy. Indexes are immutable so they can share underlying data
+        return comb_axis.copy(deep=False) if self.copy else comb_axis
```

Comment:
we *almost* never actually deep copy things. Why do you think we actually need to make this change?

Reply:
indexes are immutable and sharing them is ok.

Reply:
As mentioned by @jorisvandenbossche in #29879, it is the underlying data of indexes are considered immutable, the index object itself has some attributes (e.g. name) that can be mutated. We probably don't want changing the name of index of the concatenated object affect the original object even when ``copy=True``.

Have updated my code comment to be more precise.